aiohttp/helpers.py

```
@@ -182,6 +184,194 @@ def encode(self):
+
+    def _build_digest_header(self, method, url):
+        """
+        :rtype: str
+        """
+
+        realm = self.challenge['realm']
+        nonce = self.challenge['nonce']
+        qop = self.challenge.get('qop')
+        algorithm = self.challenge.get('algorithm', 'MD5').upper()
+        opaque = self.challenge.get('opaque')
+
+        # lambdas assume digest modules are imported at the top level
+        if algorithm == 'MD5' or algorithm == 'MD5-SESS':
+            hash_fn = hashlib.md5
+        elif algorithm == 'SHA':
+            hash_fn = hashlib.sha1
+        else:
+            return ''
+
+        def H(x):
+            return hash_fn(x.encode()).hexdigest()
+
+        def KD(s, d):
+            return H('%s:%s' % (s, d))
+
+        parsed = urlparse(url)
+        #: path is request-uri defined in RFC 2616 which should not be empty
+        path = parsed.path or "/"
+        if parsed.query:
+            path += '?' + parsed.query
+
+        A1 = '%s:%s:%s' % (self.username, realm, self.password)
+        A2 = '%s:%s' % (method, path)
+
+        HA1 = H(A1)
+        HA2 = H(A2)
+
+        if nonce == self.last_nonce:
+            self.nonce_count += 1
+        else:
+            self.nonce_count = 1
+
+        self.last_nonce = nonce
+
+        ncvalue = '%08x' % self.nonce_count
+
+        k = str(self.nonce_count).encode()
+        k += nonce.encode()
+        k += time.ctime().encode()
+        k += os.urandom(8)
+        cnonce = (hashlib.sha1(k).hexdigest()[:16])
+
+        if algorithm == 'MD5-SESS':
+            HA1 = H('%s:%s:%s' % (HA1, nonce, cnonce))
+
+        if not qop:
+            respdig = KD(HA1, '%s:%s' % (nonce, HA2))
+        elif qop == 'auth' or 'auth' in qop.split(','):
+            noncebit = '%s:%s:%s:%s:%s' % (
+                nonce, ncvalue, cnonce, 'auth', HA2
+            )
+            respdig = KD(HA1, noncebit)
+        else:
+            raise client_exceptions.ClientError(
```

`qop` is known quite early above. How about to fail fast?

---

There would have to be some duplication of logic around `qop` and I didn't think that the performance was that big of an issue. I can move it though.